### ISSUE TITLE
"No Target Architecture" when compiling libcxxwrap-julia using MSVC

### ISSUE THREAD

**Issue by barche** [JuliaLang/julia#34201]
When compiling against the Julia nightlies using MSVC, the following error appears:

```
C:\Program Files (x86)\Windows Kits\10\Include\10.0.17763.0\um\winnt.h(154,1): fatal error C1189: #error:  "No Target Architecture" [C:\projects\libcxxwrap-julia\build\cxxwrap_julia.vcxproj]
```

This comes from including `errhandlingapi.h` directly, here:
https://github.com/JuliaLang/julia/blob/2835347ff051f85997220dfe0dc152249763c6f4/src/support/dtypes.h#L351

On my local machine, I found a crude workaround, but I doubt this will work in all cases:

```c
#ifdef _OS_WINDOWS_

#define DWORD unsigned long
extern "C" __declspec(dllimport) DWORD __stdcall GetLastError();
extern "C" __declspec(dllimport) void __stdcall SetLastError(DWORD dwErrCode);

#endif
```

**Comment by barche** [comment]
I'll try to generalize my workaround above, since Julia 1.4 is now a release that has this problem and being able to use MSVC on windows is quite handy in the CxxWrap workflow for local development. Suggestions are welcome ;)

**Comment by phlavenk** [comment]
I can confirm the @barche workaround works. I'm not sure but the problems seen here are caused by the non-standard use of windows SDK headers - usually in windows programming the `errhandlingapi.h` file is never used standalone, but by including `windows.h`. I do understand that making such a portable system across so many CPU architectures , OSs and compilers is a very complex task, nicely demonstrated by the complexity of the header files, where,as I understand, many preprocessor directives normally delivered by windows.h, are workarounded directly in julia's header files. I think, as MSVC is still a supported consumer of the `julia.h` incule file, we should have some testing infrastructure in Travis that would check that the API headers are parseable by MSVC.

**Comment by nsamsonov** [comment]
Also hitting this when trying to embed julia and include `julia.h`. The code is copy-pasted from the docs:

```#include <iostream>
#include <julia.h>


int main()
{
    /* required: setup the Julia context */
    jl_init();

    /* run Julia commands */
    jl_eval_string("print(sqrt(2.0))");

    jl_atexit_hook(0);
    return 0;
}```

**Comment by rpep** [comment]
#37957 also seems to be the same issue.

**Comment by toivoh** [comment]
Thanks to an answer I found on StackOverflow, I got it work by using

    #include <uv.h>
    #include <windows.h>
    #include <julia.h>

I still get a compilation warning:

    ...\windows kits\10\include\10.0.17763.0\ucrt\signal.h(22): warning C4005: 'NSIG': macro redefinition
    ...\appdata\local\programs\julia-1.6.0\include\julia\uv\win.h(68): note: see previous definition of 'NSIG'
    
where the first definition defines `NSIG` to be 23 and the second 29, but the embedding seems to work if I also take care of https://github.com/JuliaLang/julia/issues/39595 as I described there (tried with Julia 1.6.0 in MSVC 2017).

**Comment by vtjnash** [comment]
Solved by c12aca890a8ae387d11db0b54351f8b61305c00b
